### DISCUSSION TITLE
USB Hostの利用 (RTOS)

### DISCUSSION THREAD

**Discussion by SidKoe** [KyotoMicrocomputer/solid-rapi4#4] [Q&A] [answered]
USB HostをRTOS側から使いたいと思い、少し試行してみました。

ベース 0xFE000000
オフセット 0x00980000
から、レジスタを参照しようとしたのですが、どうもうまく見えないようです。
C言語を使い、上記のアドレスを (volatile unsigned int *)のポインタに入れて、順番にアクセスすると下記の様になります。
```
0x00 00010000 00010000 00010000 00010000
0x10 00000000 00000000 00000000 00000000
0x20 90af54e1 90af54e1 90af54e1 90af54e1
0x30 00001000 00001000 00001000 00001000
```
16byte単位で同じ値が見えるようです。
参照方法に何か手順があるでしょうか？ 
(そもそも、Linux側が動作しているので、そちら側からの利用を止めるとかの必要があるのかもしれませんが)

**Comment by SidKoe** [comment]
参考までに...ここまでの間に、GIPOの操作とか、Linux側のファイル操作とかのサンプルを作ってみましたが、そちらの方は問題なく動作しています。開発環境の作成にはそれなりに問題はないのだろうと....

**Comment by takuro-tokuyama** [comment]
添付のソリューションを SOLID for RaspberryPi4 でビルド＆実行したのですが、以下のような出力になりました。
お手元の環境で実行した場合はどうなりますか？

```
[GHOST] info: Waiting for the debugger to be ready...
[GHOST] info: Calling the application entry point...
0x00 00010000 00000000 00000000 00001400
0x10 80000000 14000030 00000000 058a0820
0x20 058a0820 00001000 00201000 00080020
0x30 00000000 00000000 00000000 2708a000
```
[Project4.zip](https://github.com/KyotoMicrocomputer/solid-rapi4/files/12167649/Project4.zip)

**Reply by SidKoe** [reply]
正常に動作しています(そちらの提示された値に合致しています)。
メモリマップドIOを参照するときには、デバッガからウォッチ、メモリダンプとかをしてもうまくいかないのですね。
        システム、デバッガの構成を考えれば、ある程度予測できたような気がします。お手数おかけして申し訳ありません。
        JTAGとかで経験したままのつもりでやっていました。
変数などへの代入も試してみましたが、正常に実施できることを確認できました。
ありがとうございます。

**Reply by takuro-tokuyama** [reply, marked as answer]
ウォッチやメモリダンプでは、デバッガは対象を通常のメモリだと思ってアクセスするので、アクセスビット幅が メモリマップドIO として有効なビット幅であることが保証されません。

一度登録しておけば継続的に値を監視できるという機能では無いのですが、Partner コマンドウインドウでの、PI/PO コマンドを使用することで、明示的にアクセスビット幅を指定してメモリマップドIO のレジスタを読み書きすることができます。下記URL参照

https://solid.kmckk.com/SOLID/doc/latest/user_guide/partner_window.html

**Reply by SidKoe** [reply]
丁寧にありがとうございます。
コマンドも利用しながら試したいと思います。
